### ISSUE TITLE
dnn: wrong internal results in `expand_neg_batch` in Test_ONNX_layers.Expand

### ISSUE THREAD

**Issue by fengyuentau** [opencv/opencv#24308]
### System Information

lastest opencv and lastest onnxruntime (1.16.0, released just now)

### Detailed description

The model `expand_neg_batch.onnx` is as follows (Note that I manually added output "8".),

![image](https://github.com/opencv/opencv/assets/17219438/ad0a8b5e-3c8d-4d4c-828f-5caf821f59fa)

For output "8", onnxruntime produces value [2, 1, 1, 1]; dnn produces value [2, -1, -1, -1]. My own calculation is the same as onnxruntime's.

Model is attached here: [expand_neg_batch.new.onnx.zip](https://github.com/opencv/opencv/files/12683076/expand_neg_batch.new.onnx.zip)

---

@asmorkalov 

**Should we run every single onnx model in `test_onnx_importer.cpp` with onnxruntime and verify with our input and output data to see whether the model is correctly built and dnn can produce correct results?**

### Steps to reproduce

1. Download the model and run inference to get outputs with onnxruntime

```python
import numpy as np

import onnxruntime as ort
import cv2 as cv

x = np.random.randn(1, 2, 2).astype(np.float32)

def ort_inference(model_path):
    net = ort.InferenceSession(model_path, providers=['CPUExecutionProvider'])
    out = net.run(["8"], {})[0]
    print("ort result=", out)

def ocv_inference(model_path):
    net = cv.dnn.readNet(model_path)
    net.setInput(x)

    out = net.forward("8")
    print("ocv result=", out)

ort_inference("./expand_neg_batch.new.onnx")
# forget this, small 3d tensor is parsed incorrectly because of opencv python binding
#ocv_inference("./expand_neg_batch.new.onnx")
```

2. Clone my Expand refactor PR https://github.com/opencv/opencv/pull/24295, add the following code to print `shape` in `onnx_importer.cpp::parseExpand` (about line 2318):

```cpp
for (int i = 0; i < mat_input_shape.total(); ++i) {
    std::cout << *(mat_input_shape.ptr<int>() + i) << " ";
    // CV_Check(i, *(mat_input_shape.ptr<int>() + i) >= 0, "DNN/ONNXImporter-Expand: invalid shape dimension");
}
```

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by asmorkalov** [comment]
> Should we run every single onnx model in test_onnx_importer.cpp with onnxruntime and verify with our input and output data to see whether the model is correctly built and dnn can produce correct results?

Makes sense.

**Comment by fengyuentau** [comment]
The main issue should be addressed via #24295. The main issue is:

The wrong result is given because of `shape` is used as input shape for Expand (see line 838 and 834), which has value `[2, -1, -1, -1]` (See line 838):

https://github.com/opencv/opencv/blob/3889dcf3f816cc1b44a2b4de9cd7c5f4ebb22f70/modules/dnn/src/onnx/onnx_graph_simplifier.cpp#L824-L840

Expand does not allow neg values. We could not just use `shape` out of inference.
